### PR TITLE
Defer creating a compilation if it is not needed.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide 

We're always generating the compilation for a project in order to compute the index for FindReferences.  We do this, even if we end up not recomputing the index because nothing has changed.

**Comment by CyrusNajmabadi** [comment]
Telling @MattGertz about this.  This improves NavigateTo performance in OOP by avoiding unnecessary wasted cycles building Compilation objects.

File: src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo_Serialization.cs [hunk 1]
```diff
@@ -51,14 +46,14 @@ internal partial class SymbolTreeInfo : IObjectWritable
             Solution solution,
             Checksum checksum,
             string filePath,
-            Func<SpellChecker> create)
+            Func<Task<SpellChecker>> createAsync)
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo_Serialization.cs]: ❕ Use `Func<CancellationToken, Task<SpellChecker>>` so the scope of the token within the callback is easier to understand.

File: src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo_Serialization.cs [hunk 2]
```diff
@@ -24,22 +24,17 @@ internal partial class SymbolTreeInfo : IObjectWritable
...
-            bool loadOnly,
-            CancellationToken cancellationToken)
+            Project project, Checksum checksum, bool loadOnly, CancellationToken cancellationToken)
         {
             return LoadOrCreateAsync(
-                solution,
+                project.Solution,
                 checksum,
-                filePath,
+                project.FilePath,
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo_Serialization.cs]: What is this used for internally?  It turns out that a project's filepath is not enough to uniquely distinguish it.

**Pull request merged**